src/libpmemfile-posix/block_array.c

```diff
@@ -0,0 +1,273 @@
+allocate_new_block_array(struct pmemfile_vinode *vinode)
+{
+	TOID(struct pmemfile_block_array) new =
+			TX_ZALLOC(struct pmemfile_block_array, FILE_PAGE_SIZE);
+	D_RW(new)->length = (uint32_t)
+			((page_rounddown(pmemobj_alloc_usable_size(new.oid)) -
+			sizeof(struct pmemfile_block_array)) /
+			sizeof(struct pmemfile_block));
+
+	D_RW(new)->next = vinode->inode->file_data.blocks.next;
+	TX_SET_DIRECT(&vinode->inode->file_data.blocks, next, new);
+	vinode->first_free_block.arr = D_RW(new);
+	vinode->first_free_block.idx = 0;
+}
+
+/*
+ * is_zeroed -- check if given memory range is all zero
+ */
+static inline bool
+is_zeroed(const void *addr, size_t len)
+{
+	/* XXX optimize */
+	const char *a = (const char *)addr;
+	while (len-- > 0)
+		if (*a++)
+			return false;
+	return true;
+}
+
+static struct pmemfile_block *
+acquire_new_entry(struct pmemfile_vinode *vinode)
+{
+	ASSERT(has_free_block_entry(vinode));
+
+	struct block_info *binfo = &vinode->first_free_block;
+	struct pmemfile_block *block = binfo->arr->blocks + binfo->idx++;
+
+	ASSERT(is_zeroed(block, sizeof(*block)));
+
+	/* XXX, snapshot separated to let pmemobj use small object cache  */
+	pmemobj_tx_add_range_direct(block, 32);
+	pmemobj_tx_add_range_direct((char *)block + 32, 32);
+	COMPILE_ERROR_ON(sizeof(*block) != 64);
+
+	return block;
+}
+
+struct pmemfile_block *
+block_list_insert_after(struct pmemfile_vinode *vinode,
+			struct pmemfile_block *prev)
+{
+	update_first_block_info(vinode);
+
+	if (!has_free_block_entry(vinode))
+		allocate_new_block_array(vinode);
+
+	struct pmemfile_block *block = acquire_new_entry(vinode);
+
+	if (prev == NULL) {
+		if (vinode->first_block != NULL) {
+			block->next = blockp_as_oid(vinode->first_block);
+			TX_SET_DIRECT(vinode->first_block,
+			    prev, blockp_as_oid(block));
+		}
+		vinode->first_block = block;
+	} else {
+		block->prev = blockp_as_oid(prev);
+		block->next = prev->next;
+		TX_SET_DIRECT(prev, next, blockp_as_oid(block));
+		struct pmemfile_block *next = D_RW(block->next);
+		if (next != NULL)
+			TX_SET_DIRECT(next, prev, blockp_as_oid(block));
+	}
+
+	return block;
+}
+
+static struct pmemfile_block *
+last_used_block(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	ASSERT(binfo->idx > 0);
+
+	return binfo->arr->blocks + (binfo->idx - 1);
+}
+
+static void
+unlink_block(struct pmemfile_block *block)
+{
+	if (!TOID_IS_NULL(block->prev))
+		TX_SET(block->prev, next, block->next);
+
+	if (!TOID_IS_NULL(block->next))
+		TX_SET(block->next, prev, block->prev);
+}
+
+static void
+relocate_block(struct pmemfile_block *dst, struct pmemfile_block *src)
+{
+	ASSERT(dst != src);
+
+	TX_ADD_DIRECT(dst);
+
+	if (!TOID_IS_NULL(src->prev))
+		TX_SET(src->prev, next, blockp_as_oid(dst));
+
+	if (!TOID_IS_NULL(src->next))
+		TX_SET(src->next, prev, blockp_as_oid(dst));
+
+	TX_MEMCPY(dst, src, sizeof(*src));
+}
+
+static bool
+must_remove_a_block_array(struct pmemfile_vinode *vinode)
```

Comment:
Oops, this is bad, as the block array definite, it can be only one of the block arrays. Going to rename this, once we come up with an appropriate name.